gibbon-ghc-integration/gibbon-plugin/src/Gibbon/Plugin.hs

```diff
@@ -0,0 +1,467 @@
+import           System.Random
+import           Data.Word
+import           Data.List
+
+import qualified Gibbon.Common as Gib
+import qualified Gibbon.Pretty as Gib
+import qualified Gibbon.L0.Syntax as Gib
+import qualified Gibbon.Compiler as Gib
+import qualified Gibbon.DynFlags as Gib
+
+
+--------------------------------------------------------------------------------
+
+data Packed a = Packed (Ptr a)
+  deriving Show
+
+{-# NOINLINE liftPacked #-}
+liftPacked :: (a -> b) -> Packed a -> Packed b
+liftPacked = error "liftPacked"
+
+type PackedAnnEnv = GHC.NameEnv PackedAnn
+
+data PackedAnn = LiftPacked
+    deriving Data
+
+instance Ppr.Outputable PackedAnn where
+    ppr LiftPacked = Ppr.text "LiftPacked"
+
+
+--------------------------------------------------------------------------------
+
+
+plugin :: GHC.Plugin
+plugin = GHC.defaultPlugin { GHC.installCoreToDos = installGibbonPlugin
+                           -- , GHC.pluginRecompile = GHC.purePlugin
+                           }
+  where
+    installGibbonPlugin :: [GHC.CommandLineOption] -> [GHC.CoreToDo] -> GHC.CoreM [GHC.CoreToDo]
+    installGibbonPlugin _ todos = return (gibbonCoreTodo : todos)
+
+    gibbonCoreTodo :: GHC.CoreToDo
+    gibbonCoreTodo = GHC.CoreDoPluginPass "GibbonLiftPacked" gibbonPlugin
+
+
+gibbonPlugin :: GHC.ModGuts -> GHC.CoreM GHC.ModGuts
+gibbonPlugin mod_guts = do
+    GHC.liftIO $ print "[Gibbon Core Plugin (final)] Starting..."
+
+    -- -- (0)
+    -- GHC.putMsg (Ppr.text "\nInput module:\n----------------------------------------" Ppr.$$ (GHC.ppr (GHC.mg_binds mod_guts)))
+
+    -- (1) and (2): Find all arguments to liftPacked, form an initial root set, and then fetch the transitive closure.
+    (to_lift, dcons, closure) <- transitiveClosure mod_guts
+    GHC.putMsg (Ppr.text "\nExpressions to lift:\n----------------------------------------" Ppr.$$ (GHC.ppr (to_lift, closure)))
+
+    closure' <- mapM (\(f,rhs) ->
+                          if elem f to_lift
+                          then do
+                            let name = GHC.idName f
+                            uniq <- GHC.getUniqueM
+                            let name' = GHC.mkFCallName uniq (fromString $ "c_" ++ nameToString name)
+                            pure (GHC.setIdName f name',rhs)
+                          else pure (f,rhs))
+                     closure
+
+    -- (3) Translate the closure to Gibbon L0.
+    l0_prog <- coreToL0 dcons closure'
+    GHC.liftIO $ putStrLn "\nL0 program:\n----------------------------------------"
+    GHC.liftIO $ putStrLn (Gib.pprender l0_prog)
+
+    -- (4) Have Gibbon generate a .o file.
+    fp <- GHC.liftIO $ generateObjectFile l0_prog
+    GHC.liftIO $ putStrLn "\nCompiled:\n----------------------------------------"
+    GHC.liftIO $ putStrLn fp
+
+    -- (5) Link the .o file in the module.
+    let objfile = replaceExtension fp ".o"
+    let rtsfile = "/home/ckoparka/chai/tree-velocity/gibbon-rts/build/gibbon_rts.o"
```

Comment:
I've hardcoded various paths in a few places in the code, those all need to made relative to the "GIBBONDIR" environment variable.

Reply:
Oh, I missed this one it seems, thank you!

It's easy in the Haskell code. The issue is that some of the paths need to go into the .cabal file, where you can't rely on GIBBONDIR, I think.

Reply:
What's the best way to relativize this path, do you have an idea? My hope is that RTS can be a part of the gibbon Haskell package (as it used to be once, I think), and the plugin could depend on it (via [`extra-libraries`](https://cabal.readthedocs.io/en/3.10/cabal-package.html#pkg-field-extra-libraries) or something).

Do you think that we need to use `.o` files, or `.a` would be acceptable. Because I suppose `.a` is what Cabal will expose as an "extra-library".

Reply:
Correction: instead of `extra-libraries` read [`extra-bundled-libraries`](https://cabal.readthedocs.io/en/3.10/cabal-package.html#pkg-field-extra-bundled-libraries).

Reply:
I guess, none of the `*-libraries` fields really work because we need to recompile RTS at run time rather than compile time. Need to think more...

Reply:
I'm starting to think that since plugin can't work without Gibbon at all and it's not much code, it should be just a part of the Gibbon package. In that case it could find RTS via `data-files`. What do you think?

Reply:
Apologies, I've been having a super busy week. I just tried to merge my commit and saw this isn't a branch on this repo but on your clone. Avoiding creating a PR for a PR :D You've already cleaned up most of the things, the only additional changes I have are : (1) using GIBBONDIR in the Plugin.hs file in the spot you missed, and (2) using (isLibrary) in Codegen.hs to decide whether to codegen the main expression.

> My hope is that RTS can be a part of the gibbon Haskell package (as it used to be once, I think)

We didn't have this problem before because the old RTS was linked using copy-paste :D We had `rts.c` file that would be added as a prefix to every generated C program. 

> I guess, none of the *-libraries fields really work because we need to recompile RTS at run time rather than compile time. Need to think more...

Just to double check if I got this right, "run time" means the time that the plugin runs right? So every time a Haskell program uses `liftPacked` and the plugin kicks in? We don't need to recompile the RTS every time here. We'd want to compile the RTS just once, keep the .o/.so files in a known location, and then link the compiled Haskell program with it.

I think there might be a way avoid the cabal file altogether. We were using it to dynamically link against the `.so` file that the Rust RTS produces. Maybe there's a different way to tackle this? So the [plugin](https://github.com/iu-parfunc/gibbon/blob/wip/ghc-integration/gibbon-ghc-integration/gibbon-plugin/src/Gibbon/Plugin.hs#L99) is already inserting these .o files into the GHC compilation pipeline. Maybe we can a find a way to set rpath here too?

I didn't get a chance to try this out, but another idea was to use `Setup.hs` to set `extra-bundled-libraries:` with the right path in `gibbon-plugin.cabal` and then to only use `extra-libraries:` in the client (here `gibbon-examples.cabal`) and things should work out? Then we don't have to specify any sort of library path in the client's cabal file, because the appropriate .o/.so files should be present in the same directory as the `gibbon-plugin` library would?

> I'm starting to think that since plugin can't work without Gibbon at all and it's not much code, it should be just a part of the Gibbon package. In that case it could find RTS via data-files. What do you think?

Yeah, that's okay. But we would still need to put a relative path to the .o/.so files here or is that avoidable?

Reply:
Thanks for the reply! There are many questions, and it'd be great to untangle them. Here's an approximate list.

First, something completely orthogonal (as I realize now): I don't like that Gibbon depends on `$GIBBONDIR`. I think, as a Haskell app, it could get away with using only `Paths_gibbon`-paths. The reason I don't like the env var is that I don't see a clean way to `cabal install gibbon` in that case: you'd need to learn where the Gibbon source is stored by Cabal. This is not the way you're supposed to use Cabal :-) Before I thought that switching away from `$GIBBONDIR` might make the life of the plugin easier, but after reading your message and thinking more, I think it's probably doable in any case, so I probably need to shelve this issue for now. 

  - I thought that `data-files` will be enough to package the RTS in a reasonable way, but it seems there are catches here.
    1. You mention dynamic library for the Rust part of the RTS.
    2. Currently, the RTS is compiled "in place" meaning that the binary artifacts produced get stored under the same directory. This can break if the Gibbon package installed in a read-only filesystem. E.g. Nix store... I think the RTS artifacts should be generated in the current directory, just like the .c files for the source and alike. This is hopefully not too hard to achieve, but more work than just adding `data-files` field and switching from `$GIBBONDIR` to `Paths_gibbon` in the Haskell source of Gibbon. 

Imagine I'm compiling a plugin client with Cabal.

1. Cabal needs to find the plugin and Gibbon-the-library. This is easy is everything is packaged as one or more Haskell packages although there are several options. Gibbon and the plugin may be:
    a. two separate Haskell packages (as it is now)
    b. one package (the plugin is just a module of the Gibbon package) -- this may or may not simplify things for the rest.
    c. plugin may be a public sublibrary of the gibbon package (see [Cabal docs](https://cabal.readthedocs.io/en/3.12/cabal-package-description-file.html#pkg-field-library-visibility)) -- not sure what this could buy us (in theory, this is a cleaner interface).

3. The plugin needs to find the Gibbon RTS -- currently, we can assume it's done via `$GIBBONDIR`. In that case, it's not important what we choose in the previous step, hopefully. 
  a. Currently, the plugin assumes the RTS has been compiled it seems. That probably should be changed: I don't see why this assumption would be true in general.

4. At run time, the client needs to find the Rust part of the RTS that is compiled into a dynamic library. I am not a big fan of dynamic libraries. Can we not use a static library and deal with it in the previous step? Also, how does Gibbon-the-compiler find the `.so` now? Can't this approach be extended to the plugin somehow?..

  a.  At plugin client's run time, I hope you don't need to compile anything. When I said "run time" before, I meant plugin's run time, so the client's compile time. Anyway, I think we're on the same page about this point.

Anything I'm missing? Any comments?